packages/libs/chainweb-stream-client/src/sliding-cache.ts

```diff
@@ -0,0 +1,146 @@
+import { ITransaction } from './types';
+
+/*
+ *  A De-duplicating cache that keeps the most recent N (.cacheSpan) elements by some attribute value (.spanValueGetter callback return value)
+ *    - for cw-stream-client:
+ *      cacheSpan is the window over which we may receive duplicates when reconnecting, e.g. CONFIRMATION_DEPTH + 3
+ *      spanValueGetter returns txn => txn.height
+ *  - It must support updating elements
+ *    - for cw-stream-client: same txn can be received multiple times with different confirmation depths
+ *  - An element's identity is returned by identityCheck
+ *    - for cw-stream-client: txn => txn.meta.id
+ *  - Since elements can be updated, equality between elements with the same identity is returned by equalityCheck callback
+ *    - for cw-stream-client: txn => txn.meta.id && txn.meta.confirmations
+ *
+ *  - call .addCache([txn1, txn2, txn3])
+ *    returns bool[] like [shouldUpdate1, shouldUpdate2, shouldUpdate3]
+ *      each bool array idx corresponds to the input array element of the same idx
+ *      true iff a consumer event should be triggered, i.e. it is all of:
+ *        - within the span we are tracking, e.g. height is within [max_height - 12, max_height]
+ *        - has not seen before in its exact identity+equality configuration
+ *
+ *  Note: identity vs equality is used internally to distinguish between "push new element to cache" (no identity match found)
+ *  vs "find this item in cache and update it to the latest value (identity match found, did not equal latest version exactly)
+ *
+ */
+
+type SlidingCacheConfig<TShape extends ITransaction> = Pick<
+  SlidingCache<TShape>,
+  'spanValueGetter' | 'cacheSpan' | 'identityCheck' | 'equalityCheck'
+>;
+
+type SlidingCacheConstructor<TShape extends ITransaction> = Partial<
+  SlidingCacheConfig<TShape>
+>;
+
+const defaults: SlidingCacheConfig<ITransaction> = {
+  spanValueGetter: ({ height }) => height,
+  cacheSpan: 3,
+  identityCheck: (a, b) => a.meta?.id === b.meta?.id,
```

Comment:
Hm, good point. Since this is consuming data that came over the wire, we can't rely on TS. Two things I will fix here:

- `a?.meta?.id`
- Add an undefined/null guard in the addCache method, so we enforce we do not have unexpected non-objects